scipy/spatial/transform/rotation.py

```diff
@@ -3,12 +3,139 @@
+
+    # Step 4
+    angles = np.empty((num_rotations, 3))
+    angles[:, 1] = np.arccos(dcm_transformed[:, 2, 2])
+
+    # Steps 5, 6
+    eps = 1e-7
+    safe1 = (np.abs(angles[:, 1]) >= eps)
+    safe2 = (np.abs(angles[:, 1] - np.pi) >= eps)
+
+    # Step 4 (Completion)
+    angles[:, 1] += offset
+
+    # 5b
+    safe_mask = np.logical_and(safe1, safe2)
+    angles[safe_mask, 0] = np.arctan2(dcm_transformed[safe_mask, 0, 2],
+                                      -dcm_transformed[safe_mask, 1, 2])
+    angles[safe_mask, 2] = np.arctan2(dcm_transformed[safe_mask, 2, 0],
+                                      dcm_transformed[safe_mask, 2, 1])
+
+    if extrinsic:
+        # For extrinsic, set first angle to zero so that after reversal we
+        # ensure that third angle is zero
+        # 6a
+        angles[~safe_mask, 0] = 0
+        # 6b
+        angles[~safe1, 2] = np.arctan2(
+            dcm_transformed[~safe1, 1, 0] - dcm_transformed[~safe1, 0, 1],
+            dcm_transformed[~safe1, 0, 0] + dcm_transformed[~safe1, 1, 1]
+        )
+        # 6c
+        angles[~safe2, 2] = -np.arctan2(
+            dcm_transformed[~safe2, 1, 0] + dcm_transformed[~safe2, 0, 1],
+            dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+        )
+    else:
+        # For instrinsic, set third angle to zero
+        # 6a
+        angles[~safe_mask, 2] = 0
+        # 6b
+        angles[~safe1, 0] = np.arctan2(
+            dcm_transformed[~safe1, 1, 0] - dcm_transformed[~safe1, 0, 1],
+            dcm_transformed[~safe1, 0, 0] + dcm_transformed[~safe1, 1, 1]
+        )
+        # 6c
+        angles[~safe2, 0] = np.arctan2(
+            dcm_transformed[~safe2, 1, 0] + dcm_transformed[~safe2, 0, 1],
+            dcm_transformed[~safe2, 0, 0] - dcm_transformed[~safe2, 1, 1]
+        )
+
+    # Step 7
+    if seq[0] == seq[2]:
+        # lambda = 0, so we can only ensure angle2 -> [0, pi]
+        adjust_mask = np.logical_or(angles[:, 1] < 0, angles[:, 1] > np.pi)
+    else:
+        # lambda = + or - pi/2, so we can ensure angle2 -> [-pi/2, pi/2]
+        adjust_mask = np.logical_or(angles[:, 1] < -np.pi / 2,
+                                    angles[:, 1] > np.pi / 2)
+
+    # Dont adjust gimbal locked angle sequences
+    adjust_mask = np.logical_and(adjust_mask, safe_mask)
+
+    angles[adjust_mask, 0] += np.pi
+    angles[adjust_mask, 1] = 2 * offset - angles[adjust_mask, 1]
+    angles[adjust_mask, 2] -= np.pi
+
+    angles[angles < -np.pi] += 2 * np.pi
+    angles[angles > np.pi] -= 2 * np.pi
+
+    # Step 8
+    if not np.all(safe_mask):
+        warnings.warn("Gimbal lock detected. Setting third angle to zero since"
+                      " it is not possible to uniquely determine all angles.")
+
+    # Reverse role of extrinsic and intrinsic rotations, but let third angle be
+    # zero for gimbal locked cases
+    if extrinsic:
+        angles = angles[:, ::-1]
+    return angles
+
+
+def make_elementary_quat(axis, angles):
```

Comment:
Make private

I'll stop commenting on each of these, but I assume you get the idea.